core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexUpgradeService.java

```diff
@@ -83,13 +93,25 @@ public MetaDataIndexUpgradeService(Settings settings) {
      */
     public IndexMetaData upgradeIndexMetaData(IndexMetaData indexMetaData) {
         // Throws an exception if there are too-old segments:
+        if(isUpgraded(indexMetaData)) {
+            return indexMetaData;
+        }
         checkSupportedVersion(indexMetaData);
         IndexMetaData newMetaData = upgradeLegacyRoutingSettings(indexMetaData);
         newMetaData = addDefaultUnitsIfNeeded(newMetaData);
+        newMetaData = upgradeMappings(newMetaData);
+        newMetaData = markAsUpgraded(newMetaData);
         return newMetaData;
     }
 
     /**
+     * Checks if the index was already opened by this version of Elasticsearch and doesn't require any additional checks.
+     */
+    private boolean isUpgraded(IndexMetaData indexMetaData) {
+        return indexMetaData.creationVersion().onOrAfter(Version.V_2_0_0_beta1) || indexMetaData.upgradeVersion().onOrAfter(Version.V_2_0_0_beta1);
```

Comment:
If upgrade version is not present, it is set to created version, so we can just check upgrade version.